src/color.rs

```diff
@@ -1,54 +1,71 @@
 use num_traits::{NumCast, Zero};
-use std::mem;
 use std::ops::{Index, IndexMut};
 
 use buffer::Pixel;
 use traits::Primitive;
 
-/// An enumeration over supported color types and their bit depths
+/// An enumeration over supported color types and bit depths
 #[derive(Copy, PartialEq, Eq, Debug, Clone, Hash)]
 pub enum ColorType {
-    /// Pixel is grayscale
-    Gray(u8),
-
-    /// Pixel contains R, G and B channels
-    RGB(u8),
-
-    /// Pixel is an index into a color palette
-    Palette(u8),
-
-    /// Pixel is grayscale with an alpha channel
-    GrayA(u8),
-
-    /// Pixel is RGB with an alpha channel
-    RGBA(u8),
-
-    /// Pixel contains B, G and R channels
-    BGR(u8),
-
-    /// Pixel is BGR with an alpha channel
-    BGRA(u8),
-
+    /// Pixel is 1-bit luminance
+    L1,
+
+    /// Pixel is 8-bit luminance
+    L8,
+    /// Pixel is 8-bit luminance with an alpha channel
+    LA,
+    /// Pixel contains 8-bit R, G and B channels
+    RGB,
+    /// Pixel is 8-bit RGB with an alpha channel
+    RGBA,
+
+    /// Pixel is 16-bit luminance
+    L16,
+    /// Pixel is 16-bit luminance with an alpha channel
+    LA16,
+    /// Pixel is 16-bit RGB
+    RGB16,
+    /// Pixel is 16-bit RGBA
+    RGBA16,
+
+    /// Pixel contains 8-bit B, G and R channels
+    BGR,
+    /// Pixel is 8-bit BGR with an alpha channel
+    BGRA,
+
+    /// Pixel is of unknown color type with the specified bit depth. This can apply to pixels which
+    /// are associated with an external palette. In that case, the pixel value is an index into the
+    /// palette.
+    Unknown(u8),
```

Comment:
I'm not sure if this is a sufficient 'catch all' type for what it looks like it is trying to achieve. On the one hand it is still too open to have somewhat guaranteed support (no encoder in its right mind has 255-bit depth support) but too restricted to abstract color planes. I don't think we should restrict the representation here, `Error::Unsupported` can be returned where necessary. So I'd rather embrace it as (or similar):

```
Unknown {
    planes: u8,
    bits_per_plane: u8,
},
```

Reply:
But as a pure escape hatch with out-of-band information it's kind of weird. The decoder needs to return byte data. The maximum flexibility would then be it returning simply how many bytes-per-pixel are required since it can not expect the further processing steps to have any idea about the correct transformation for the color information. Thus `bit depth` is almost meaningless to the environment.

What happens if we just drop the variant entirely? Decoders still have the option of returning an `Unsupported` error for some sort.

Reply:
I am also in favor of dropping `Unknown` and replacing it with some sort of `Unsupported` error, and simply remove all the `From` impls that can fail. We have the argument that `Unknown` really cannot be used to do anything which is pretty strong. 

A simple way to do this could probably be with a `#[doc(hidden)]` enum variant `__Unknown`. Since the enum already is unexhaustive any place where it is used needs some sort of wildcards which should return some `Err(Unsupported(...))`.

Reply:
Maybe that was not such a good idea, it would make the enum quite a bit more painful to use.

Reply:
Would it be clearer if the description was changed from `bit depth` to `bits per pixel`? This is the meaning I've been intending. If we removed the case entirely, we'd have a worse regression until proper decoding of RGBA2 and similar formats are supported.

Reply:
Maybe `bytes per pixel` would be a feasible middle ground. All internal loading post-processing expands pixels to byte boundaries, i.e. `L1` is turned to `u8`. But for arbitrary `bits` there is no clear process to do this.

Reply:
I'm not sure all decoders do the post processing you are describing. From my understanding, the PNG decoder doesn't do any expansion for BitDepth::{One, Two, Four} and so as a result RGB images will have bits per pixels that aren't multiples of 8.

Reply:
If that conversion is actually happening that is a bigger bug: we're claiming the returned image data will be in one format and then providing it in a different one

Reply:
Afaik, it does: https://github.com/image-rs/image-png/blob/4f2e8eb44586442133c0222e47e1eb762aaef5a5/src/decoder/mod.rs#L304-L311 and https://github.com/image-rs/image-png/blob/4f2e8eb44586442133c0222e47e1eb762aaef5a5/src/decoder/mod.rs#L349-L355 and `Transformations::EXPAND` is set by default. 

I'm not sure if I consider that a bug by the way. The format is that of the *returned buffer* which is `L8` since we don't support others. It doesn't seem too weird to me that you don't get the genuine format of the *image* itself through that same interface, that seems by definition like it exceeds the possibilities of the limited number of enum variants.

Reply:
Does this mean that L1 is also impossible to get from decoders? If we can guarantee that pixels will always be an integer number of bytes then I think it does make sense to have Unknown bytes_per_pixel and to remove L1.

Regarding 2-bit RGB, 4-bit palette, and such the PNG crate actually has enums for those so I (wrongly?) assumed they were supported not just for input but also for the format of returned buffers. If we can just match over the results of png::Reader::output_color_type() and panic unimplemented!() on any sub 8-bit formats that would make the code much nicer.

Reply:
This really boils down to what `ColorType` is supposed to represent, as according to the discussion in #855 after is RFC was opened. Either

* Collection of known color formats in the wild (basically what it was previously but less wildly unstructured).
* The color formats from which we can always convert to one of the formats of images. Then `Unknown` doesn't quite fit except if we introduce an `Opaque` pixel data representation. But `L4` should then be surely included as well.
* The color formats that the image buffers can *actually* have, in particular those on which generic algorithms can operate. I think that's too restrictive.

One design goal could be: Be liberal in what to expect and restrictive in what to emit. That would point to create *two* color types namely one for the each of the last two options. Sounds like a hassle but that might actually solve a few problems. There is no buffer representation utilizing `L4` but a roundtrip of `png` encoded with it should definitely be possible. Two different color types would make it possible to store the original encoding independent of the color type of the buffer and offer an interface that accepts the former while providing a default if only the latter is provided (as in `DynamicImage::write_to`).

Reply:
The `ColorType` is currently used to mean (and I'd argue should continue to mean):

* The set of color formats that decoders can chose for their output data or that users can given to encoders as input. Thus `L4` and similar aren't needed since the decoder can always "decode" that to `L8` (and that is easier for the user because they don't have to deal with weird formats)

In #873 I previously tried to create a separate enum for (3), but discovered that it would be used almost nowhere since both encoders and decoders want a less restricted set of formats and the Pixel trait / DynamicImage control which formats generic algorithms operate over.

I don't have strong feelings on `Unknown` vs `Opaque`, they mean roughly the same thing to me.

Reply:
Your definition seems phrases somewhere between my `(2)` and `(3)` but I fail to see the distinction between requiring the decoder to already convert to `L8` and requiring it to convert to the set of colors that `DynamicImage` is supposed to represent. I.e it sounds exactly like (3) but the paragraph below says that it isn't supposed to.

Reply:
`DynamicImage` doesn't support ColorType::{L, LA, RGB, RGBA}16, though I suppose that may also change at some point. For now I see two main directions forward:

1. We can merge things as they are now, perhaps changing Unknown -> Opaque first.

2. We can remove L1, make Unknown/Opaque take *bytes* per pixel, and require at all future colortypes have pixels which are an integer number of bytes. Any decoders which produce output less than 8 bpp would be converted to return an error (and then a future PR could do an expand pass instead).

Reply:
One thing to consider with regards to `Unknown`: due to the reference semantics we will have to require that any `impl Pixel` type is an integer number of bytes in any case (meaning `L1` will *never* be represented as an `image::Pixel` anytime soon). Since unknown can not be converted (including 'scaling') it *can never* be stored in a `GenericImage`-like buffer if a pixel is not composed of a whole number of bytes. I don't think it would be prospective to suggest otherwise.

Why I think it should be two types: The primary concern of `DynamicImage` is to ensure that the data is represented in some `Pixel` implementation (to allow it to implement `GenericImage`). However `L1` and `L4` and several other formats can not be a pixel as it conflates two samples in a single byte and that won't work with references. That means, in particular, that `ColorType` does not seem to be intended as a discriminant of a `Pixel`-type enum. On an image with such a source color type there has to, necessarily, be some post-processing.

I agree it would be nice however *not* to lose that information. And I also think that for some color types the transformation should be done by this library when such post-processing has a clearly defined meaning. That avoids redundancy and makes decoders easier to write. *Be lenient in what to accept*. But then why stop at `L1` and not allow `L4`? The conversion is reasonably clear in both directions. What about `Yuv420`? Wouldn't it be nice to have the `image` library automatically duplicate the `cr` channel into each pixel's sample?

As laid out above, it would also be inconvenient to have `Pixel::color_type` defined in terms of that same enum due to the integer-byte-count etc. In fact, the mapping of a `ColorType` to representations in pixel buffers would be non-injective (and thus can not be reversed). And due to L16 as observed by you it might error. This question would be avoided when having two enums. One for the original output of the decoder and one for the resulting representation as a `GenericImage`. Perhaps name one `ColorType` and one `PixelType`.

It would also (possibly) solve the problem for encoders. Since it is encouraged not to do the preprocessing on its own the input color format information is preserved – only the image buffer into which it is loaded doesn't have the information on its own but it can be queried from the decoder beforehand in standardized types. Then, the encoder can be supplied with both during encoding or the data be pre-processed to that 

So, about the resolution of this. If we split this into `ColorType` and `PixelType` we have no more problems. The `DynamicImage` code would try to find a `PixelType` that the `ColorType` returned by the decoder can be converted into, then do that conversion if possible. In reverse, the `write_to` implementation would find a default `ColorType` for the `PixelType` (possibly being wasteful compared to the original encoding) but another, new api could offer supplying a `ColorType` directly and convert the data back before feeding it into the encoder. And with regards to `ColorType::Unknown(bytes)`; the `DynamicImage` loading would fail with an error of `Unsupported` but reading directly would still work. This seems to be the most flexible option where required. We could then also expand on the variants of `ColorType` to offer more decoders with zero-loss of information while maintaining the `PixelType` of their output images (e.g. introduce `Rgb555` but the output image would still be `RGB8`).

Reply:
Tl;dr: Basically option 1. but expanded right away to not lose functionality?

Reply:
That approach seems reasonable. Any opposition to merging this now and then adding `PixelType` in a follow up?